File: shared/WaiSayKit/Sources/WaiSayKit/Models/Recording.swift
import Foundation

/// Recording types
public enum RecordingType: String, Codable, Sendable, CaseIterable {
    case meeting
    case note
    case reflection
}

/// Recording upload / processing lifecycle.
public enum RecordingStatus: String, Codable, Sendable, CaseIterable {
    case pendingUpload = "pending_upload"
    case uploading
    case processing
    case ready
    case failed

    public var label: String {
        switch self {
        case .pendingUpload:
            return "Waiting to Sync"
        case .uploading:
            return "Syncing in Background"
        case .processing:
            return "Processing"
        case .ready:
            return "Ready"
        case .failed:
            return "Needs Attention"
        }
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let rawStatus = try container.decode(String.self)
        guard let status = RecordingStatus(rawValue: rawStatus) else {
            throw DecodingError.dataCorruptedError(
                in: container,
                debugDescription: "Unknown RecordingStatus value: '\(rawStatus)'"
            )
        }
        self = status
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(rawValue)
    }
}

/// Folder for organizing recordings.
public struct Folder: Codable, Identifiable, Sendable {
    public let id: String
    public var name: String
    public let createdAt: Date

    public init(id: String, name: String, createdAt: Date = Date()) {
        self.id = id
        self.name = name
        self.createdAt = createdAt
    }

    private enum CodingKeys: String, CodingKey {
        case id
        case name
        case createdAt = "created_at"
    }
}

/// Public share link returned when a recording owner creates a web share.
public struct RecordingShareLink: Codable, Sendable {
    public let recordingId: String
    public let token: String
    public let url: URL
    public let createdAt: Date

    public init(recordingId: String, token: String, url: URL, createdAt: Date = Date()) {
        self.recordingId = recordingId
        self.token = token
        self.url = url
        self.createdAt = createdAt
    }

    private enum CodingKeys: String, CodingKey {
        case recordingId = "recording_id"
        case token
        case url
        case createdAt = "created_at"
    }
}

/// Recording model
public struct Recording: Codable, Identifiable, Sendable {
    public let id: String
    public var title: String?
    public let type: RecordingType
    public let audioUrl: String?
    public let status: RecordingStatus
    public let failureCode: String?
    public let failureMessage: String?
    public let uploadedAt: Date?
    public let durationSeconds: Int?
    public let language: String?
    public let folderId: String?
    public let deletedAt: Date?
    public let starredAt: Date?
    public let createdAt: Date

    public init(
        id: String,
        title: String? = nil,
        type: RecordingType,
        audioUrl: String? = nil,
        status: RecordingStatus = .pendingUpload,
        failureCode: String? = nil,
        failureMessage: String? = nil,
        uploadedAt: Date? = nil,
        durationSeconds: Int? = nil,
        language: String? = nil,
        folderId: String? = nil,
        deletedAt: Date? = nil,
        starredAt: Date? = nil,
        createdAt: Date = Date()
    ) {
        self.id = id
        self.title = title
        self.type = type
        self.audioUrl = audioUrl
        self.status = status
        self.failureCode = failureCode
        self.failureMessage = failureMessage
        self.uploadedAt = uploadedAt
        self.durationSeconds = durationSeconds
        self.language = language
        self.folderId = folderId
        self.deletedAt = deletedAt
        self.starredAt = starredAt
        self.createdAt = createdAt
    }

    private enum CodingKeys: String, CodingKey {
        case id
        case title
        case type
        case audioUrl = "audio_url"
        case status
        case failureCode = "failure_code"
        case failureMessage = "failure_message"
        case uploadedAt = "uploaded_at"
        case durationSeconds = "duration_seconds"
        case language
        case folderId = "folder_id"
        case deletedAt = "deleted_at"
        case starredAt = "starred_at"
        case createdAt = "created_at"
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        title = try container.decodeIfPresent(String.self, forKey: .title)
        type = try container.decode(RecordingType.self, forKey: .type)
        audioUrl = try container.decodeIfPresent(String.self, forKey: .audioUrl)
        let decodedStatus = try container.decodeIfPresent(RecordingStatus.self, forKey: .status)
        status = decodedStatus ?? .pendingUpload
        failureCode = try container.decodeIfPresent(String.self, forKey: .failureCode)
        failureMessage = try container.decodeIfPresent(String.self, forKey: .failureMessage)
        uploadedAt = try container.decodeIfPresent(Date.self, forKey: .uploadedAt)
        durationSeconds = try container.decodeIfPresent(Int.self, forKey: .durationSeconds)
        language = try container.decodeIfPresent(String.self, forKey: .language)
        folderId = try container.decodeIfPresent(String.self, forKey: .folderId)
        deletedAt = try container.decodeIfPresent(Date.self, forKey: .deletedAt)
        starredAt = try container.decodeIfPresent(Date.self, forKey: .starredAt)
        createdAt = try container.decode(Date.self, forKey: .createdAt)
    }
}

/// A key moment / highlight from a recording.
public struct RecordingHighlight: Codable, Identifiable, Sendable {
    public let id: String
    public let recordingId: String
    public let category: String
    public let title: String
    public let description: String?
    public let speaker: String?
    public let startMs: Int?
    public let endMs: Int?
    public let importance: String

    private enum CodingKeys: String, CodingKey {
        case id
        case recordingId = "recording_id"
        case category
        case title
        case description
        case speaker
        case startMs = "start_ms"
        case endMs = "end_ms"
        case importance
    }
}

/// Recording with full details including segments and summary
public struct RecordingDetail: Codable, Identifiable, Sendable {
    public let id: String
    public var title: String?
    public let type: RecordingType
    public let audioUrl: String?
    public let status: RecordingStatus
    public let failureCode: String?
    public let failureMessage: String?
    public let uploadedAt: Date?
    public let durationSeconds: Int?
    public let language: String?
    public let folderId: String?
    public let deletedAt: Date?
    public let starredAt: Date?
    public let createdAt: Date
    public let segments: [Segment]
    public let summary: Summary?
    public let actionItems: [ActionItem]
    public let highlights: [RecordingHighlight]

    public init(
        id: String,
        title: String? = nil,
        type: RecordingType,
        audioUrl: String? = nil,
        status: RecordingStatus = .pendingUpload,
        failureCode: String? = nil,
        failureMessage: String? = nil,
        uploadedAt: Date? = nil,
        durationSeconds: Int? = nil,
        language: String? = nil,
        folderId: String? = nil,
        deletedAt: Date? = nil,
        starredAt: Date? = nil,
        createdAt: Date = Date(),
        segments: [Segment] = [],
        summary: Summary? = nil,
        actionItems: [ActionItem] = [],
        highlights: [RecordingHighlight] = []
    ) {
        self.id = id
        self.title = title
        self.type = type
        self.audioUrl = audioUrl
        self.status = status
        self.failureCode = failureCode
        self.failureMessage = failureMessage
        self.uploadedAt = uploadedAt
        self.durationSeconds = durationSeconds
        self.language = language
        self.folderId = folderId
        self.deletedAt = deletedAt
        self.starredAt = starredAt
        self.createdAt = createdAt
        self.segments = segments
        self.summary = summary
        self.actionItems = actionItems
        self.highlights = highlights
    }

    private enum CodingKeys: String, CodingKey {
        case id
        case title
        case type
        case audioUrl = "audio_url"
        case status
        case failureCode = "failure_code"
        case failureMessage = "failure_message"
        case uploadedAt = "uploaded_at"
        case durationSeconds = "duration_seconds"
        case language
        case folderId = "folder_id"
        case deletedAt = "deleted_at"
        case starredAt = "starred_at"
        case createdAt = "created_at"
        case segments
        case summary
        case actionItems = "action_items"
        case highlights
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        title = try container.decodeIfPresent(String.self, forKey: .title)
        type = try container.decode(RecordingType.self, forKey: .type)
        audioUrl = try container.decodeIfPresent(String.self, forKey: .audioUrl)
        let decodedStatus = try container.decodeIfPresent(RecordingStatus.self, forKey: .status)
        status = decodedStatus ?? .pendingUpload
        failureCode = try container.decodeIfPresent(String.self, forKey: .failureCode)
        failureMessage = try container.decodeIfPresent(String.self, forKey: .failureMessage)
        uploadedAt = try container.decodeIfPresent(Date.self, forKey: .uploadedAt)
        durationSeconds = try container.decodeIfPresent(Int.self, forKey: .durationSeconds)
        language = try container.decodeIfPresent(String.self, forKey: .language)
        folderId = try container.decodeIfPresent(String.self, forKey: .folderId)
        deletedAt = try container.decodeIfPresent(Date.self, forKey: .deletedAt)
        starredAt = try container.decodeIfPresent(Date.self, forKey: .starredAt)
        createdAt = try container.decode(Date.self, forKey: .createdAt)
        segments = try container.decodeIfPresent([Segment].self, forKey: .segments) ?? []
        summary = try container.decodeIfPresent(Summary.self, forKey: .summary)
        actionItems = try container.decodeIfPresent([ActionItem].self, forKey: .actionItems) ?? []
        highlights = try container.decodeIfPresent([RecordingHighlight].self, forKey: .highlights) ?? []
    }
}

/// Request to create a new recording
public struct CreateRecordingRequest: Codable, Sendable {
    public var title: String?
    public var type: RecordingType
    public var language: String
    public var folderId: String?

    public init(
        title: String? = nil,
        type: RecordingType = .note,
        language: String = "en",
        folderId: String? = nil
    ) {
        self.title = title
        self.type = type
        self.language = language
        self.folderId = folderId
    }

    private enum CodingKeys: String, CodingKey {
        case title
        case type
        case language
        case folderId = "folder_id"
    }
}

public extension Recording {
    var isFailedUpload: Bool {
        status == .failed
    }

    var failurePreviewText: String? {
        UserFacingErrorFormatter.previewMessage(failureMessage, context: .recording)
    }

    func statusDisplayText(hasLocalRecoveryBackup: Bool = false) -> String? {
        if hasLocalRecoveryBackup && !isFailedUpload {
            return "Saved locally"
        }

        switch status {
        case .failed:
            return "Needs attention"
        case .pendingUpload:
            return "Waiting to sync"
        case .uploading:
            return "Syncing in background"
        case .processing:
            return "Processing"
        case .ready:
            return nil
        }
    }

    var statusDisplayText: String? {
        statusDisplayText()
    }
}

File: backend/app/models/recording.py
"""Recording and related models."""

import enum
import uuid
from datetime import date, datetime, timezone
from typing import TYPE_CHECKING

from pgvector.sqlalchemy import Vector
from sqlalchemy import (
    Boolean,
    Date,
    DateTime,
    Float,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    false,
    text,
)
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin, UUIDMixin

if TYPE_CHECKING:
    from app.models.person import Person


class RecordingStatus(str, enum.Enum):
    """Lifecycle states for recording upload and processing."""

    PENDING_UPLOAD = "pending_upload"
    UPLOADING = "uploading"
    PROCESSING = "processing"
    READY = "ready"
    FAILED = "failed"


ACTIVE_RECORDING_STATUSES = frozenset(
    {
        RecordingStatus.PENDING_UPLOAD.value,
        RecordingStatus.UPLOADING.value,
        RecordingStatus.PROCESSING.value,
    }
)


class SummaryGenerationStatus(str, enum.Enum):
    """Lifecycle states for manual summary generation jobs."""

    QUEUED = "queued"
    RUNNING = "running"
    SUCCEEDED = "succeeded"
    FAILED = "failed"


class Folder(Base, UUIDMixin, TimestampMixin):
    """Folder for organizing recordings."""

    __tablename__ = "folders"
    # Note: unique constraint on (user_id, name) deferred — migration 000010 removed
    # due to production deploy issues. Will re-add with proper pre-validation.

    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)

    # Relationships
    user: Mapped["User"] = relationship("User", back_populates="folders")
    recordings: Mapped[list["Recording"]] = relationship("Recording", back_populates="folder")


class Recording(Base, UUIDMixin, TimestampMixin):
    """Audio recording model."""

    __tablename__ = "recordings"
    # RETURNING-fetch server defaults (e.g. updated_at onupdate) on flush so the
    # ORM object isn't left with an expired attribute that a sync serializer would
    # try to lazy-load (MissingGreenlet) after an UPDATE.
    __mapper_args__ = {"eager_defaults": True}

    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    title: Mapped[str | None] = mapped_column(String(500))
    # True while the title is auto-generated (provisional or summary-derived); a
    # user rename flips it False so the authoritative summary title never clobbers
    # a manual one. See apply_summary_result + update_recording.
    title_auto_generated: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=True,
        server_default=text("true"),
    )
    type: Mapped[str] = mapped_column(String(50), nullable=False)  # meeting, note, reflection
    audio_url: Mapped[str | None] = mapped_column(String(1000))
    status: Mapped[str] = mapped_column(
        String(32),
        nullable=False,
        default=RecordingStatus.PENDING_UPLOAD.value,
    )
    failure_code: Mapped[str | None] = mapped_column(String(100))
    failure_message: Mapped[str | None] = mapped_column(Text)
    uploaded_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    duration_seconds: Mapped[int | None] = mapped_column(Integer)
    billed_word_count: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=0,
        server_default="0",
    )
    language: Mapped[str | None] = mapped_column(String(10))
    folder_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("folders.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    deleted_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
        index=True,
    )
    starred_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )
    # Trust/salience signals for the P4 ranking multiplier (neutral defaults).
    authority_score: Mapped[float] = mapped_column(
        Float, nullable=False, server_default="0.5", default=0.5
    )
    salience_score: Mapped[float | None] = mapped_column(Float)

    # Relationships
    user: Mapped["User"] = relationship("User", back_populates="recordings")
    folder: Mapped["Folder | None"] = relationship("Folder", back_populates="recordings")
    segments: Mapped[list["Segment"]] = relationship(
        "Segment", back_populates="recording", cascade="all, delete-orphan"
    )
    summary: Mapped["Summary | None"] = relationship(
        "Summary", back_populates="recording", cascade="all, delete-orphan", uselist=False
    )
    summary_generation_jobs: Mapped[list["SummaryGenerationJob"]] = relationship(
        "SummaryGenerationJob",
        back_populates="recording",
        cascade="all, delete-orphan",
        order_by="desc(SummaryGenerationJob.created_at)",
    )
    summary_audio_artifacts: Mapped[list["SummaryAudioArtifact"]] = relationship(
        "SummaryAudioArtifact",
        back_populates="recording",
        cascade="all, delete-orphan",
        order_by="desc(SummaryAudioArtifact.created_at)",
    )
    action_items: Mapped[list["ActionItem"]] = relationship(
        "ActionItem", back_populates="recording", cascade="all, delete-orphan"
    )
    tags: Mapped[list["RecordingTag"]] = relationship(
        "RecordingTag", back_populates="recording", cascade="all, delete-orphan"
    )
    highlights: Mapped[list["Highlight"]] = relationship(
        "Highlight", back_populates="recording", cascade="all, delete-orphan"
    )
    share_links: Mapped[list["RecordingShare"]] = relationship(
        "RecordingShare", back_populates="recording", cascade="all, delete-orphan"
    )


class RecordingShare(Base, UUIDMixin, TimestampMixin):
    """Public share link for a recording.

    Only a SHA-256 hash of the bearer token is stored. The raw token is returned
    once when the owner creates a share link and is later verified by hashing the
    token from the public URL.
    """

    __tablename__ = "recording_shares"

    recording_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("recordings.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    token_hash: Mapped[str] = mapped_column(String(64), nullable=False, unique=True, index=True)
    revoked_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
        index=True,
    )

    recording: Mapped["Recording"] = relationship("Recording", back_populates="share_links")


class Segment(Base, UUIDMixin):
    """Transcript segment with speaker diarization."""

    __tablename__ = "segments"

    recording_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("recordings.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    speaker: Mapped[str | None] = mapped_column(String(100))
    raw_label: Mapped[str | None] = mapped_column(String(100))
    person_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("people.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    auto_assigned: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=False, server_default=false()
    )
    match_confidence: Mapped[float | None] = mapped_column(Float)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    start_ms: Mapped[int | None] = mapped_column(Integer)
    end_ms: Mapped[int | None] = mapped_column(Integer)
    confidence: Mapped[float | None] = mapped_column(Float)
    embedding: Mapped[list[float] | None] = mapped_column(Vector(1536))

    # Relationships
    recording: Mapped["Recording"] = relationship("Recording", back_populates="segments")
    person: Mapped["Person | None"] = relationship("Person")


class Summary(Base, UUIDMixin, TimestampMixin):
    """AI-generated summary of a recording."""

    __tablename__ = "summaries"

    recording_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("recordings.id", ondelete="CASCADE"),
        nullable=False,
        unique=True,
    )
    summary: Mapped[str | None] = mapped_column(Text)
    key_points: Mapped[list | None] = mapped_column(JSONB)
    decisions: Mapped[list | None] = mapped_column(JSONB)
    topics: Mapped[list | None] = mapped_column(JSONB)
    people_mentioned: Mapped[list | None] = mapped_column(JSONB)
    sentiment: Mapped[str | None] = mapped_column(String(20))

    # Relationships
    recording: Mapped["Recording"] = relationship("Recording", back_populates="summary")


class SummaryGenerationJob(Base, UUIDMixin, TimestampMixin):
    """Durable server-side state for manual summary generation."""

    __tablename__ = "summary_generation_jobs"
    __table_args__ = (
        Index("ix_summary_generation_jobs_recording_id", "recording_id"),
        Index("ix_summary_generation_jobs_user_id", "user_id"),
        Index("ix_summary_generation_jobs_status", "status"),
        Index(
            "ux_summary_generation_jobs_active_recording",
            "recording_id",
            unique=True,
            postgresql_where=text("status IN ('queued', 'running')"),
        ),
    )

    recording_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("recordings.id", ondelete="CASCADE"),
        nullable=False,
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )
    status: Mapped[str] = mapped_column(
        String(32),
        nullable=False,
        default=SummaryGenerationStatus.QUEUED.value,
        server_default=SummaryGenerationStatus.QUEUED.value,
    )
    stage: Mapped[str] = mapped_column(
        String(64),
        nullable=False,
        default="queued",
        server_default="queued",
    )
    progress_percent: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=5,
        server_default="5",
    )
    transcript_hash: Mapped[str] = mapped_column(String(64), nullable=False)
    instructions_override: Mapped[str | None] = mapped_column(Text)
    task_id: Mapped[str | None] = mapped_column(String(255))
    error_code: Mapped[str | None] = mapped_column(String(100))
    error_message: Mapped[str | None] = mapped_column(Text)
    requested_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    failed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    attempt_count: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=0,
        server_default="0",
    )

    recording: Mapped["Recording"] = relationship(
        "Recording",
        back_populates="summary_generation_jobs",
    )
    user: Mapped["User"] = relationship("User")


class ActionItem(Base, UUIDMixin, TimestampMixin):
    """Action item extracted from a recording."""

    __tablename__ = "action_items"

    recording_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("recordings.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    task: Mapped[str] = mapped_column(Text, nullable=False)
    owner: Mapped[str | None] = mapped_column(String(200))
    due_date: Mapped[date | None] = mapped_column(Date)
    priority: Mapped[str | None] = mapped_column(String(20))  # high, medium, low
    status: Mapped[str] = mapped_column(String(20), default="pending")
    source: Mapped[str] = mapped_column(String(20), default="generated", nullable=False)

    # Relationships
    recording: Mapped["Recording"] = relationship("Recording", back_populates="action_items")


# Import at bottom to avoid circular imports
from app.models.entity import RecordingTag  # noqa: E402
from app.models.highlight import Highlight  # noqa: E402
from app.models.summary_audio import SummaryAudioArtifact  # noqa: E402
from app.models.user import User  # noqa: E402
